compiler/src/iree/compiler/Dialect/HAL/Analysis/Attributes/DeviceTargetPVS.cpp

```diff
@@ -0,0 +1,250 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/HAL/Analysis/Attributes/DeviceTargetPVS.h"
+
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "iree/compiler/Dialect/Util/IR/UtilTypes.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "iree-hal-device-analysis"
+
+namespace mlir::iree_compiler::IREE::HAL {
+
+//===----------------------------------------------------------------------===//
+// DeviceTargetGlobalPVS
+//===----------------------------------------------------------------------===//
+
+const char DeviceTargetGlobalPVS::ID = 0;
+
+void DeviceTargetGlobalPVS::initializeOperation(IREE::Util::GlobalOp globalOp,
+                                                DFX::Solver &solver) {
+  assert(isa<IREE::HAL::DeviceType>(globalOp.getType()) &&
+         "only initialize on globals of type !hal.device");
+
+  // We only support immutable initialized device globals.
+  // We could track usage up through stores to handle the mutable case but
+  // the compiler does not generate such programs today.
+  auto *globalInfo = solver.getExplorer().getGlobalInfo(globalOp);
```

Comment:
My thinking is that we'll have cases where we want to adopt devices or pass devices across module boundaries. E.g. a top-level module acting as a pipeline/the application could pick the device and pass the handle in to a lower-level module, or a low-level module that is device-specific could pick a vmfb compiled for the specific device and a higher-level module (e.g. pipeline) could be compiled to work with many and just inherit that device. I believe the latter should work today if the higher-level modules are always compiled with a superset of the devices in the low-level ones, but the former needs mutable globals as the global would be passed into a set_device method or something that shares the device instance instead of relying on enumeration to select the same device from the list of available devices.

May be a YAGNI, but hopefully we get more pipelines authored in torch/mlir/etc and this becomes a normal working mode :)